### ISSUE TITLE
Make the Scala runtime independent of the standard library

### ISSUE THREAD

**Issue by odersky** [lampepfl/dotty-feature-requests#15]
It would be a nice achievement if "vanilla"(*) compiler-generated code was not dependent on the full standard library. Benefits include:

 - easier interop with other languages,
 - possibly smaller footprint,
 - less upgrade and migration headaches due to better separation of concerns,
 - easier to use alternative libraries.

This would require at least the following changes:

 - Create a small, self-contained kernel library. This would have to contain tuple and function types,
    StringContext, Option, what else? We need to refactor these classes to have no external
    dependencies to the standard library at large.
 - Add an immutable array type to the runtime library, which erases to standard arrays, but is
    immutable.
 - Make vararg parameters take an immutable array instead of a Seq. This has the potential of
   simplifying Java interop at the price of complicating interop with 2.13.
 - Revise the compiler-generated Enum methods to not use Seqs or Maps, but to use immutable
   arrays instead.

(*) "Vanilla" means that there are some parts of the language that will require rich library support.  E.g. a quoted expression will necessarly pull in the full tasty library. But it would be good to find a natural subset of the language that is guaranteed to work only with the kernel library.

Questions: Is this feasible? Are there issues I have overlooked so far? Should we make the effort?

**Comment by performantdata** [comment]
Does [this](https://gist.github.com/performantdata/5f3f719abcf88459763cea267d6f727b) (begin) to help? It's just the superclass (black) and interfaces (blue) dependencies for now.

**Comment by bblfish** [comment]
I have a use case for this feature which could help provide an extra motivation on top of architectural elegance. 

I am working on helping a project developing a [Categorical Database transformation tool](https://www.categoricaldata.net/) based on work Functorial Databases (Databases understood as functors from small categories to Set). They produce a jar that is 14MB in size. 

I needed to write a bit of Java to parse RDF Nodes (URIs and Literals) that can be loaded from the class path. Initially I took some code from [banana-rdf](https://github.com/banana-rdf/banana-rdf) and quickly got a very flexible prototype that way.   I tried packaging it in one jar to put it in the class path and found that jar to be 35MB in size, which is ok for a prototype, but a bit problematic if there is a bug (which there was) as it makes it difficult to tell if the bug is in my code or theirs. It is also a bit awkward to have what was meant to be a plugin be twice as large as the product. That was not going to help me convince them to allow me to use Scala to program...

So I decided to extract the parser from banana-rdf adapt it a bit and compile it. I had written it to use only types like Int and Char for efficiency, and all I had to do was remove some of the nice features of Scala I had used. The aim was to not have to depend on the scala libraries as that still adds 5MB jar.  So I wrote up a question on Scala Users [Writing Java in Scala](https://users.scala-lang.org/t/writing-java-in-scala/6374), and got some good very helpful feedback pointing me also to this issue.

In the end I got the code to produce a 14K jar. But I can't easily tell if I still have some dependency on the Scala library (though Boxing for example). 

In short three are use cases for 1. Testing 2. Interacting with Java and 3. having a good opening when introducing Scala to a team using Java 4. Allowing people like me to not have to remember the muscle memory of Java syntax.
